IdentityCore/src/cache/MSIDTokenCacheKey.m

```diff
@@ -169,4 +156,135 @@ + (NSString *)familyClientId:(NSString *)familyId
+
+- (BOOL)isEqual:(id)object
+{
+    if (self == object)
+    {
+        return YES;
+    }
+    
+    if (![object isKindOfClass:MSIDTokenCacheKey.class])
+    {
+        return NO;
+    }
+    
+    return [self isEqualToTokenCacheKey:(MSIDTokenCacheKey *)object];
+}
+
+- (NSUInteger)hash
+{
+    NSUInteger hash = 17;
+    hash = hash * 31 + self.account.hash;
+    hash = hash * 31 + self.service.hash;
+    hash = hash * 31 + self.type.hash;
+    
+    return hash;
+}
+
+#pragma mark - Private
+
++ (NSString *)serviceWithAuthority:(NSURL *)authority
+                          resource:(NSString *)resource
+                          clientId:(NSString *)clientId
+{
+    
+    return [NSString stringWithFormat:@"%@|%@|%@|%@",
+            s_adalLibraryString,
+            authority.absoluteString.msidBase64UrlEncode,
+            [self.class getAttributeName:resource.msidBase64UrlEncode],
+            clientId.msidBase64UrlEncode];
+}
+
++ (NSString *)serviceWithAuthority:(NSURL *)authority
+                            scopes:(NSOrderedSet<NSString *> *)scopes
+                          clientId:(NSString *)clientId
+{
+    if (scopes.count == 0)
+    {
+        return nil;
+    }
+    
+    return [NSString stringWithFormat:@"%@$%@$%@",
+            authority? authority.absoluteString.msidBase64UrlEncode : @"",
+            clientId? clientId.msidBase64UrlEncode : @"",
+            scopes? scopes.msidToString.msidBase64UrlEncode : @""];
+}
+
+#pragma mark - NSCopying
+
+- (id)copyWithZone:(NSZone *)zone
+{
+    MSIDTokenCacheKey *key = [[MSIDTokenCacheKey allocWithZone:zone] init];
+    key.account = [self.account copyWithZone:zone];
+    key.service = [self.service copyWithZone:zone];
+    key.type = [self.type copyWithZone:zone];
+    key.authority = [self.authority copyWithZone:zone];
+    key.resource = [self.resource copyWithZone:zone];
+    key.clientId = [self.clientId copyWithZone:zone];
+    
+    return key;
+}
+
+#pragma mark - NSSecureCoding
+
++ (BOOL)supportsSecureCoding
+{
+    return YES;
+}
+
+- (instancetype)initWithCoder:(NSCoder *)coder
+{
+    if (!(self = [super init]))
+    {
+        return nil;
+    }
+    
+    _account = [coder decodeObjectOfClass:[NSString class] forKey:@"account"];
+    _service = [coder decodeObjectOfClass:[NSString class] forKey:@"service"];
+    _type = [coder decodeObjectOfClass:[NSNumber class] forKey:@"type"];
+    
+    NSString *authority = [coder decodeObjectOfClass:[NSString class] forKey:@"authority"];
+    if (authority)
+    {
+        _authority = [[NSURL alloc] initWithString:authority];
+    }
+    
+    _resource = [coder decodeObjectOfClass:[NSString class] forKey:@"resource"];
+    _clientId = [coder decodeObjectOfClass:[NSString class] forKey:@"clientId"];
+    
+    if (!_service)
+    {
+        _service = [self.class serviceWithAuthority:self.authority resource:self.resource clientId:self.clientId];
+    }
+    
+    return self;
+}
+
+- (void)encodeWithCoder:(NSCoder *)coder
```

Comment:
should it be JSON serializable too?

Reply:
We don't serialize it to json (we use only 'service' & 'account' properties).

Reply:
We don't serialize it to JSON now, we might in future. Please add a Github issue to track this.

Reply:
added #55